### DISCUSSION TITLE
Diagnosing a type guard failure

### DISCUSSION THREAD

**Discussion by jatoben** [microsoft/pyright#3748] [Q&A]
I have some code which is performing a geometry operation that looks essentially like this:

```python
# source_point is a Point
# dest_target is a Point | None

if dest_target is not None:
    target = dest_target
else:
    target = source_point

dest_edge, dest_point = min(
    ((e, e.center()) for e in dest_edges),
    key=lambda ep: geometry.dist(ep[1], target),
)
```

Pyright 1.1.263 reports that `target` (parameter `p2`) might be `None`, although that shouldn't be possible:

```
error: Argument of type "Point | None" cannot be assigned to parameter "p2" of type "Point | Shift" in function "dist"
    Type "Point | None" cannot be assigned to type "Point | Shift"
      Type "None" cannot be assigned to type "Point | Shift"
        Type "None" cannot be assigned to type "Point"
        Type "None" cannot be assigned to type "Shift" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
Completed in 0.622sec
```

Interestingly, if I add a `reveal_type(source_point)` before the `target` assignment (but not after), Pyright correctly reports that `source_point` is a `Point`, not `Point | None`, and the error is silenced.

```
information: Type of "source_point" is "Point"
0 errors, 0 warnings, 1 information 
Completed in 0.684sec
```

Adding only a `reveal_type(target)` after the `target` assignment shows that `target` is a `Point`, but does not quiet the error.

```
error: Argument of type "Point | None" cannot be assigned to parameter "p2" of type "Point | Shift" in function "dist"
    Type "Point | None" cannot be assigned to type "Point | Shift"
      Type "None" cannot be assigned to type "Point | Shift"
        Type "None" cannot be assigned to type "Point"
        Type "None" cannot be assigned to type "Shift" (reportGeneralTypeIssues)
information: Type of "target" is "Point"
1 error, 0 warnings, 1 information 
Completed in 0.632sec
```

I'm not sure if I'm encountering a bug or if I'm missing something, and I haven't yet had success in extracting this bit of code into a minimal test case. Any tips on tracking it down? Thanks!

**Comment by erictraut** [comment]
I don't think I can diagnose without a self-contained minimal example. Your code above references a bunch of symbols that are not declared or imported.

**Reply by jatoben** [reply]
Thanks @erictraut. I'll keep working on isolating a smaller test case.
